File: examples/example1.cpp
// Copyright (c) 2025 Matheus Machado Fiuza <matheusmachadofiuza@gmail.com>

#include "numav/numav.hpp"

int main()
{
    using namespace numav;

    // create the simulation object with some numerical method
    auto s = Simulation<
        Phenomenon::ACOUSTIC,
        NumericalMethod::FEM,
        Domain::FREQUENCY,
        Dimension::D3,
        ElementOrder::O1
    >();

    // load the mesh
    s.load_mesh("build/tests_bin/test1.bdf");

    // determine simulation frequency range
    double freq_min = 20;
    double freq_max = 100;
    s.set_frequency_range(freq_min, freq_max);

    s.add_volume_material(
        1,
        [](auto f) { return 1.2; },
        [](auto f) { return std::complex<double>(343,0); }
    );
    s.add_volume_material(
        2,
        [](auto f) { return 1.2; },
        [](auto f) { return std::complex<double>(343,5); }
    );

    // add volume velocity sources
    auto q = [](auto f) { return 1/f; };
    s.add_sound_source(
        TypeOfSource::POINT, {3.0, 2.0, 1.0},
        PhysicalQuantity::VOLUME_VELOCITY, q
    );
    auto u = [](auto f) { return 1/f; };
    s.add_sound_source(
        TypeOfSource::SURFACE, 2,
        PhysicalQuantity::PARTICLE_VELOCITY, u
    );

    // add pressure sources
    auto p0 = [](auto f) { return 1; };
    s.add_sound_source(
        TypeOfSource::POINT, {4.0, 1.0, 0.5},
        PhysicalQuantity::PRESSURE, p0
    );
    auto p1 = [](auto f) { return 2; };
    s.add_sound_source(
        TypeOfSource::SURFACE, 3,
        PhysicalQuantity::PRESSURE, p1
    );
 
    // add specific surface acoustic impedance
    auto Z = [](auto f) { return std::complex<double>(4000, 4000); };
    s.add_surface_specific_acoustic_impedance(5, Z);

    // run the simulation
    s.run();

    // export the result to binary
    s.export_result("result.nmvr");
}

File: examples/example1.jl
# Copyright (c) 2025 Matheus Machado Fiuza <matheusmachadofiuza@gmail.com>

using Numav

# create the simulation object with some numerical method
s = Simulation{
    Phenomenon.acoustic,
    NumericalMethod.fem,
    Domain.frequency,
    Dimension.d3
}()

# set element order
set_element_order(s, 2)

# determine simulation frequency range
freq_min = 0
freq_max = 100
set_freq_limits(s, freq_min, freq_max)

# load the mesh
load_mesh(s, "example1.bdf")

# add a volume material (air in this case)
id_air = 1
rho_air = 1.20
c_air = 343
add_volume_material(s, id_air, rho_air, c_air)

# add volume velocity sources
Q(f) = 10/f
add_source(s, coordinates=[1.0, 1.5, 2.0], volume_velocity=Q)
add_source(s, surface_id=2, volume_velocity=Q)

# add pressure sources
P(f) = 2/f
add_source(s, coordinates=[2.0, 2.5, 1.0], pressure=P)
add_source(s, surface_id=3, pressure=P)

# add specific surface impedance
Z(f) = 1f + 2im
add_specific_surface_acoustic_impedance(s, surface_id=4, impedance=Z)

# run the simulation
result = Numav.run(s)
